### PR TITLE
refactor: process pegouts at end of epoch

### PR DESCRIPTION
**Pull request by scottmillner**
Testing:
I have tested this and stepped through the code and verified it's working up to sending the pegout request to `btc_server` which then prints the following error: `"Invalid tx request, dkg has not been completed"`
I have no reason to believe the pegout request will not be successful once DKG is fully implemented

Overview:

- When node spins up, it stores all pegouts in the epoch in `InnerStorage`
- When node is building the last block in the epoch, it gets pegouts from the cache and sends them to `btc_server`
- The node clears all pegouts from the cache once above is done

Open questions:

- What do we do with failed pegouts?
- Current implementation loops through pegouts and just logs the ones that fail and continues processing all pegouts until finished
- The failed pegouts don't carry over into the next epoch and are lost funds for the user

### CODE REVIEW

**Review by arminsabouri — COMMENTED** [review]
Awesome work!!! Left some questions and comments.

File: crates/consensus/authority/src/block_builder.rs [hunk 1]
```diff
@@ -221,9 +203,25 @@ where
...
+                info!(target: "consensus::authority", "Sending pegouts: {:?}", pegouts);
+                match send_pegouts(&self.bitcoin_block_source, &mut self.btc_server, pegouts).await
+                {
+                    Ok(_) => {}
+                    Err(e) => {
+                        error!(target: "consensus::authority", ?e, "Failed to send all pegouts");
+                    }
+                }
+
+                storage.pegouts.clear();
```
Comment (arminsabouri) [inline crates/consensus/authority/src/block_builder.rs]: The problem here is that regarless of error or not the cache gets cleared. If the signing session restarts, signers have no way of validating the pegouts.

File: crates/consensus/authority/src/utils.rs [hunk 2]
```diff
@@ -192,31 +226,20 @@ async fn process_botanix_log(
...
-
-                let raw_tx = response.into_inner().tx;
-                info!(target: "consensus::authority", "broadcasting withdrawal tx");
-
-                bitcoin_block_source
-                    .broadcast_tx(&hex::encode(raw_tx))
-                    .await
-                    .map_err(|_| ProcessBotanixLogError::FailedToBroadcastPegout)?;
+                info!(target: "consensus::authority", "Adding pegout to cache");
+                storage.write().await.pegouts.push(request);
```
Comment (arminsabouri) [inline crates/consensus/authority/src/utils.rs]: What happens here in the case of a re-org? This is a downside of the cache I think.

File: crates/consensus/authority/src/block_builder.rs (hunk 1 above)
Comment (scottmillner) [reply crates/consensus/authority/src/block_builder.rs]: per our conversation will remove pegouts from cache and just grab all at the end of epoch and process

File: crates/consensus/authority/src/utils.rs (hunk 2 above)
Comment (scottmillner) [reply crates/consensus/authority/src/utils.rs]: removing pegouts from cache

**Comment by scottmillner** [comment]
superseded by https://github.com/botanix-labs/Macbeth/pull/112

**Pull request closed without merging**